File: tests/fixtures/cs-strategy-pos/Pricing.cs
using System;

namespace App;

// SMELL: pricing logic branches on a `kind` string with >=3 cases. The same
// if/else cascade is conceptually duplicated wherever a price is computed
// (checkout + invoice). Candidate for Strategy (see references/patterns/strategy.md).
public static class Pricing
{
    public static double Price(string kind, double @base)
    {
        if (kind == "regular")
        {
            return @base;
        }
        if (kind == "vip")
        {
            return @base * 0.8;
        }
        if (kind == "staff")
        {
            return @base * 0.5;
        }
        throw new ArgumentException("unknown kind: " + kind);
    }
}

File: tests/fixtures/kotlin-strategy-neg/app/src/main/java/app/Pricing.kt
package app

// SMELL: same switch over kind repeated in multiple sites — Strategy candidate.
fun price(kind: String, base: Double): Double {
    if (kind == "regular") return base
    if (kind == "vip") return base * 0.8
    if (kind == "staff") return base * 0.5
    error("unknown kind: $kind")
}

fun discountedTotal(kind: String, base: Double, qty: Int): Double {
    val unit: Double
    if (kind == "regular") unit = base
    else if (kind == "vip") unit = base * 0.8
    else if (kind == "staff") unit = base * 0.5
    else error("unknown kind: $kind")
    return unit * qty
}
